modules/facts/plans/retrieve.pp

```diff
@@ -0,0 +1,46 @@
+# A plan that retrieves facts from the specified nodes by running
+# an apropriate facts::* task on each. Care is taken to run a
+# facts::* task corresponding to the node's platfrom on each
+# node. If there is no variant of the task appropriate for some
+# node's platform the node is treated as if the task had failed.
+#
+# The $nodes parameter is a list of the nodes to retrieve the facts
+# from.
+plan facts::retrieve(TargetSpec $nodes) {
+  $targets = get_targets($nodes)
+
+  # Build a mapping from the names of the tasks to run to the lists of
+  # targets to run the tasks on
+  $task_targets = $targets.facts::group_by |$target| {
+    $target.protocol ? {
+      'ssh'   => 'facts::bash',
+      'winrm' => 'facts::powershell',
+      'pcp'   => 'facts::ruby',
+      'local' => if (facts::is_command_available('bash')) { 'facts::bash' } else { undef },
```

Comment:
If you have local in your plan and don't have bash there are probably bigger problems at this point.

Reply:
That's probably true. Drop `is_command_available`?

Reply:
it seems like it removes a lot of complexity we can just let the normal error bubble up.